packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -74,324 +186,476 @@ class FocusNode extends ChangeNotifier {
+  /// this node, if it's a scope.
+  ///
+  /// Returns null if no scope is found.
+  ///
+  /// Use [enclosingScope] if you want to only look for scopes above this node.
+  FocusScopeNode get nearestScope => enclosingScope;
+
+  /// Returns the nearest enclosing scope node above this node, or null if none
+  /// is found.
+  ///
+  /// If this node is itself a scope, this will only return ancestors of this
+  /// scope.
+  ///
+  /// Use [nearestScope] to start at this node instead of above it.
+  FocusScopeNode get enclosingScope {
+    return ancestors.firstWhere((FocusNode node) => node is FocusScopeNode, orElse: () => null);
+  }
+
+  // Sets this node as the focused child for the enclosing scope, and that scope
+  // as focused child for the scope above it, etc., until it reaches the root
+  // node.
+  void _setAsFocusedChild() {
+    FocusNode scopeFocus = this;
+    for (FocusScopeNode ancestor in ancestors.whereType<FocusScopeNode>()) {
+      assert(scopeFocus != ancestor, "Somehow made a loop by setting focusedChild to it's scope.");
+      // Remove it anywhere in the focused children history.
+      ancestor._focusedChildren.remove(scopeFocus);
+      // And add it to the back (the top of the queue, and so the current one).
+      ancestor._focusedChildren.add(scopeFocus);
+      scopeFocus = ancestor;
+    }
+  }
+
+  /// Returns the size of the associated [Focus] in logical units.
+  Size get size => context.findRenderObject().semanticBounds.size;
+
+  /// Returns the global offset to the upper left corner of the [Focus] in
+  /// logical units.
+  Offset get offset {
+    final RenderObject object = context.findRenderObject();
+    return MatrixUtils.transformPoint(object.getTransformTo(null), object.semanticBounds.topLeft);
+  }
+
+  /// Returns the global rectangle surrounding the node in logical units.
+  Rect get rect {
+    final RenderObject object = context.findRenderObject();
+    final Offset globalOffset = MatrixUtils.transformPoint(object.getTransformTo(null), object.semanticBounds.topLeft);
+    return globalOffset & object.semanticBounds.size;
   }
 
-  void _resignFocus(FocusNode node) {
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<BuildContext>('context', context, defaultValue: null));
+    properties.add(FlagProperty('hasFocus', value: hasFocus, ifTrue: 'FOCUSED', defaultValue: false));
+    properties.add(StringProperty('debugLabel', debugLabel, defaultValue: null));
+  }
+
+  @override
+  List<DiagnosticsNode> debugDescribeChildren() {
+    final List<DiagnosticsNode> children = <DiagnosticsNode>[];
+    int count = 1;
+    for (FocusNode child in _children) {
+      children.add(child.toDiagnosticsNode(name: 'child $count'));
+      ++count;
+    }
+    return children;
+  }
+}
+
+/// A [FocusNode] subclass that acts as a scope for its descendants,
```

Comment:
This probably bears a longer explanation. Why do FocusScopeNodes exist, i.e. why would I want to create one? Are there existing widgets that create them?